core/parachain/validator/parachain_processor.hpp

```diff
@@ -7,39 +7,299 @@
 #define KAGOME_PARACHAIN_PROCESSOR_HPP
 
 #include <memory>
+#include <queue>
+#include <thread>
+#include <unordered_map>
 
+#include <boost/asio/executor_work_guard.hpp>
+#include <boost/asio/io_context.hpp>
+#include <boost/asio/io_service.hpp>
+#include <boost/asio/post.hpp>
+#include <boost/asio/signal_set.hpp>
+#include <libp2p/peer/peer_id.hpp>
+
+#include "common/visitor.hpp"
+#include "crypto/hasher.hpp"
 #include "network/protocols/req_collation_protocol.hpp"
 #include "network/types/collator_messages.hpp"
+#include "outcome/outcome.hpp"
+#include "primitives/common.hpp"
 #include "utils/non_copyable.hpp"
+#include "utils/safe_object.hpp"
 
 namespace kagome::network {
   class PeerManager;
   class Router;
+  struct PendingCollation;
 }  // namespace kagome::network
 
 namespace kagome::crypto {
   class Sr25519Provider;
-}
+  class Hasher;
+}  // namespace kagome::crypto
 
 namespace kagome::parachain {
 
-  struct ParachainProcessorImpl {
+  struct ParachainProcessorImpl
+      : std::enable_shared_from_this<ParachainProcessorImpl> {
+    enum class Error {
+      RESPONSE_ALREADY_RECEIVED = 1,
+      COLLATION_NOT_FOUND,
+      KEY_NOT_PRESENT
+    };
+    static constexpr uint64_t kBackgroundWorkers = 5;
+
+    struct ImportStatementSummary {
+      /// The digest of the candidate.
+      primitives::BlockHash candidate;
+      /// The group that the candidate is in.
+      network::ParachainId group_id;
+      /// How many validity votes are currently witnessed.
+      uint64_t validity_votes;
+      /// Attested more than threshold
+      bool attested;
+    };
+
     ParachainProcessorImpl(
         std::shared_ptr<network::PeerManager> pm,
         std::shared_ptr<crypto::Sr25519Provider> crypto_provider,
-        std::shared_ptr<network::Router> router);
-    ~ParachainProcessorImpl() = default;
+        std::shared_ptr<network::Router> router,
+        std::shared_ptr<boost::asio::io_context> this_context,
+        std::shared_ptr<crypto::Sr25519Keypair> keypair,
+        std::shared_ptr<crypto::Hasher> hasher);
+    ~ParachainProcessorImpl();
 
-    void requestCollations();
+    void start();
+    void stop();
+    void requestCollations(network::PendingCollation const &pending_collation);
+    void setAssignedParachain(std::optional<network::ParachainId> para_id);
+    void handleStatement(libp2p::peer::PeerId const &peer_id,
+                         primitives::BlockHash const &relay_parent,
+                         std::shared_ptr<network::Statement> const &statement);
 
    private:
+    enum struct CollationState { kFetched, kSeconded };
+    enum struct PoVDataState { kComplete, kFetchFromValidator };
+    enum struct WorkersState : uint64_t { kWork = 0, kExit };
+    enum struct BackgroundOperationType {
```

Comment:
Current single usage can be replaced with lambda/function, or member function pointer